src/stan/mcmc/hmc/xhmc/base_xhmc.hpp

```diff
@@ -0,0 +1,223 @@
+      }
+
+      int get_max_depth() { return this->max_depth_; }
+      double get_max_deltaH() { return this->max_deltaH_; }
+      double get_x_delta() { return this->x_delta_; }
+
+      sample transition(sample& init_sample,
+                        interface_callbacks::writer::base_writer& writer) {
+        // Initialize the algorithm
+        this->sample_stepsize();
+
+        this->seed(init_sample.cont_params());
+
+        this->hamiltonian_.sample_p(this->z_, this->rand_int_);
+        this->hamiltonian_.init(this->z_, writer);
+
+        ps_point z_plus(this->z_);
+        ps_point z_minus(z_plus);
+
+        ps_point z_sample(z_plus);
+        ps_point z_propose(z_plus);
+
+        double sum_numer = this->hamiltonian_.dG_dt(this->z_);
+        double sum_weight = 1;
+
+        double H0 = this->hamiltonian_.H(this->z_);
+        int n_leapfrog = 0;
+        double sum_metro_prob = 1;  // exp(H0 - H0)
+
+        // Build a trajectory until the NUTS criterion is no longer satisfied
+        this->depth_ = 0;
+        this->divergent_ = 0;
+
+        while (this->depth_ < this->max_depth_) {
+          // Build a new subtree in a random direction
+          bool valid_subtree = false;
+          double sum_numer_subtree = 0;
+          double sum_weight_subtree = 0;
+
+          if (this->rand_uniform_() > 0.5) {
+            this->z_.ps_point::operator=(z_plus);
+            valid_subtree
+              = build_tree(this->depth_, z_propose,
+                           sum_numer_subtree, sum_weight_subtree,
+                           H0, 1, n_leapfrog, sum_metro_prob, writer);
+            z_plus.ps_point::operator=(this->z_);
+          } else {
+            this->z_.ps_point::operator=(z_minus);
+            valid_subtree
+              = build_tree(this->depth_, z_propose,
+                           sum_numer_subtree, sum_weight_subtree,
+                           H0, -1, n_leapfrog, sum_metro_prob, writer);
+            z_minus.ps_point::operator=(this->z_);
+          }
+
+          sum_numer += sum_numer_subtree;
+          sum_weight += sum_weight_subtree;
+          if (!valid_subtree) break;
+
+          // Sample from an accepted subtree
+          ++(this->depth_);
+
+          double accept_prob = sum_weight_subtree / sum_weight;
+          if (this->rand_uniform_() < accept_prob)
+            z_sample = z_propose;
+
+            // Break if exhaustion criterion is satisfied
+            if (std::fabs(sum_numer / sum_weight) < x_delta_)
+              break;
+        }
+
+        this->n_leapfrog_ = n_leapfrog;
+
+        // Compute average acceptance probabilty across entire trajectory,
+        // even over subtrees that may have been rejected
+        double accept_prob
+          = sum_metro_prob / static_cast<double>(n_leapfrog + 1);
+
+        this->z_.ps_point::operator=(z_sample);
+        this->energy_ = this->hamiltonian_.H(this->z_);
+        return sample(this->z_.q, -this->z_.V, accept_prob);
+      }
+
```

Comment:
This is really awesome.  The algorithm's as clean (if not cleaner) in C++ than in the paper's pseudocode!